### ISSUE TITLE
Handling in/out params

### ISSUE THREAD

**Issue by muglug** [vimeo/psalm#1088]
Given a function
```php
function addFoo(?string &$s) : void {
  if ($s === null) {
    $s = "hello";
  }
  $s .= "foo";
}
function takesString(string $_) : void {}

$a = null;
addFoo($a); // $a is typed to string|null

takesString($a); // fails with PossiblyNullArgument
```
I want to have some way of annotating what happens to the type.

Currently Psalm assumes that whatever type comes out is the same as the one coming in, and prohibits [changing that type inside the function](https://getpsalm.org/r/ca94124954).

But it would be great to instead have an annotation like

```php
/**
 * @param-out string $s
 */
function addFoo(?string &$s) : void {
  if ($s === null) {
    $s = "hello";
  }
  $s .= "foo";
}
```

You could then combine this with templates for more exotic docblocks e.g.
```php
/**
 * @template TKey
 * @template TValue
 * @template TNewValue
 *
 * @param TNewValue $x
 * @param array<TKey, TValue> $arr
 * @param-out array<TKey|int, TValue|TNewValue>
 */
function array_push(array &$arr, ...$x) : int {}
```

cc @TysonAndre @ondrejmirtes @JanTvrdik

**Comment by TysonAndre** [comment]
Phan has `@phan-output-reference` as a modifier for `@param` - https://github.com/phan/phan/wiki/Annotating-Your-Source-Code#phan-output-reference

I think that'd be equivalent to `@param-in mixed $var`, `@param-out SomeType $var`

The proposal sounds reasonable, but I don't have support for `@template` in functions/methods yet, so it'd be a while before this gets implemented in Phan.

**Comment by muglug** [comment]
since `@param` and `@param-in` are the same from PHP's perspective, I'll just add `@param-out`

**Comment by muglug** [comment]
Fixed in ee2b06a, though Psalm doesn't _verify_ that the method returns the out param yet
